File: 8. String to Integer (atoi)/String to Integer (atoi).java
class Solution {
    public int myAtoi(String str) {
        int num = 0;
        int sign = 1;
        int i = 0;
        int n = str.length();
        while(i < n && str.charAt(i) == ' ') i++;
        if(i < n && str.charAt(i) == '+'){
            i++;
        } else if(i < n && str.charAt(i) == '-'){
            sign = -1;
            i++;
        }
        for(; i < n; i++){
            if(str.charAt(i) < '0' || str.charAt(i) > '9') break;
            if(num > Integer.MAX_VALUE / 10 || (num == Integer.MAX_VALUE / 10 && (str.charAt(i) - '0') > Integer.MAX_VALUE % 10)){
                return sign == -1 ? Integer.MIN_VALUE : Integer.MAX_VALUE;
            }
            num = num * 10 + (str.charAt(i) - '0');
        }
        return sign * num;
    }
}

File: 8. String to Integer (atoi)/String to Integer (atoi).py
class Solution(object):
    def myAtoi(self, str):
        """
        :type str: str
        :rtype: int
        """
        
        ls = list(str.strip())
        if len(ls) == 0: return 0
        sign = -1 if ls[0] == '-' else 1
        if ls[0] in ['+', '-']: del ls[0]
        ans, i = 0, 0
        while i < len(ls) and ls[i].isdigit():
            ans = ans * 10 + ord(ls[i]) - ord('0')
            i += 1
        if sign * ans > 2**31 -1:
            return 2**31 - 1
        elif sign * ans < -2**31:
            return -2**31
        else: return sign * ans
